components/pd_client/src/client_v2.rs

```diff
@@ -0,0 +1,1412 @@
+                force,
+                ctx.enable_forwarding,
+                false,
+            )
+            .await
+        {
+            Err(e) => {
+                PD_RECONNECT_COUNTER_VEC
+                    .with_label_values(&["failure"])
+                    .inc();
+                return Err(e);
+            }
+            Ok(None) => {
+                PD_RECONNECT_COUNTER_VEC
+                    .with_label_values(&["no-need"])
+                    .inc();
+                return Ok(false);
+            }
+            Ok(Some(tuple)) => {
+                PD_RECONNECT_COUNTER_VEC
+                    .with_label_values(&["success"])
+                    .inc();
+                tuple
+            }
+        };
+
+        fail_point!("pd_client_v2_reconnect", |_| Ok(true));
+
+        self.channel = channel;
+        self.stub = stub;
+        self.target_info = target_info;
+        self.members = members;
+        info!("trying to update PD client done"; "spend" => ?start.saturating_elapsed());
+        Ok(true)
+    }
+}
+
+#[derive(Clone)]
+pub struct CachedRawClient {
+    context: Arc<ConnectContext>,
+    latest: Arc<Mutex<RawClient>>,
+    version: Arc<AtomicU64>,
+    cache: RawClient,
+    cache_version: u64,
+}
+
+impl CachedRawClient {
+    fn new(context: ConnectContext, client: RawClient) -> Self {
+        Self {
+            context: Arc::new(context),
+            latest: Arc::new(Mutex::new(client.clone())),
+            version: Arc::new(AtomicU64::new(0)),
+            cache: client,
+            cache_version: 0,
+        }
+    }
+
+    #[inline]
+    async fn refresh_cache(&mut self) -> bool {
+        if self.cache_version < self.version.load(Ordering::Acquire) {
+            let latest = self.latest.lock().await;
+            self.cache = (*latest).clone();
+            self.cache_version = self.version.load(Ordering::Relaxed);
+            true
+        } else {
+            false
+        }
+    }
+
+    /// Refreshes the local cache with latest client, then waits for the
+    /// connection to be ready.
+    async fn wait_for_ready(&mut self) -> Result<()> {
+        self.refresh_cache().await;
+        for _ in 0..3 {
+            if self
+                .channel()
+                .wait_for_connected(Duration::from_secs(REQUEST_TIMEOUT))
+                .await
+            {
+                return Ok(());
+            } else if !self.refresh_cache().await {
+                // Only retry if client is updated.
+                break;
+            }
+        }
+        Err(box_err!(
+            "Connection unavailable {:?}",
+            self.channel().check_connectivity_state(false)
+        ))
+    }
+
+    /// Increases global version only when a new connection is established.
+    async fn reconnect(&mut self, force: bool, callback: &CallbackHolder) -> Result<()> {
+        let mut latest = self.latest.lock().await;
+        if latest.maybe_reconnect(&self.context, force).await? {
```

Comment:
If there is a single background corountine to manage the lifetime of client/channel, then it doesn't need to reconnect within lock.

Reply:
When it's reconnect, it means the current version does not work anyway,  so it does no hurt to have a lock here?

Reply:
Other API callers may be blocked for arbitrary duration.